cirq-core/cirq/experiments/readout_confusion_matrix.py

```diff
@@ -0,0 +1,351 @@
+
+        Raises:
+            ValueError: If `qubits` is not a subset of `self.qubits`.
+        """
+
+        if qubits is None:
+            qubits = self.qubits
+        if any(q not in self.qubits for q in qubits):
+            raise ValueError(f"qubits {qubits} should be a subset of self.qubits {self.qubits}.")
+        return np.linalg.pinv(self.confusion_matrix(qubits))
+
+    def apply(
+        self,
+        result: np.ndarray,
+        qubits: Optional[Sequence['cirq.Qid']] = None,
+        *,
+        method='least_squares',
+    ) -> np.ndarray:
+        """Applies corrections to the observed `result` to compensate for readout error on qubits.
+
+        The compensation can applied by the following methods:
+         1. 'pseudo_inverse': The result is multiplied by the correction matrix, which is pseudo
+                              inverse of confusion matrix corresponding to the subspace defined by
+                              `qubits`.
+         2. 'least_squares': Solves a constrained minimization problem to find optimal `x` s.t.
+                                a) x >= 0
+                                b) sum(x) == sum(result) and
+                                c) sum((result - x @ confusion_matrix) ** 2) is minimized.
+
+        Args:
+            result: `(2 ** len(qubits), )` shaped numpy array containing observed frequencies /
+                    probabilities.
+            qubits: Sequence of qubits used for sampling to get `result`. By default, uses all
+                    qubits in sorted order, i.e. `self.qubits`. Note that ordering of qubits sets
+                    the basis ordering for the `result` argument.
+            method: Correction Method. Should be either 'pseudo_inverse' or 'least_squares'.
+                    Equal to `least_squares` by default.
+
+        Returns:
+              `(2 ** len(qubits), )` shaped numpy array corresponding to `result` with corrections.
+
+        Raises:
+            ValueError: if `result.shape` != `(2 ** len(qubits),)`.
+        """
+        if qubits is None:
+            qubits = self.qubits
+        if result.shape != (2 ** len(qubits),):
+            raise ValueError(f"result.shape {result.shape} should be {(2 ** len(qubits),)}.")
+        if method not in ['pseudo_inverse', 'least_squares']:
+            raise ValueError(f"method: {method} should be 'pseudo_inverse' or 'least_squares'.")
+
+        if method == 'pseudo_inverse':
+            return result @ self.correction_matrix(qubits)  # coverage: ignore
+
+        # Least squares minimization.
+        cm = self.confusion_matrix(qubits)
+
+        def func(x):
+            return np.sum((result - x @ cm) ** 2)
+
+        constraints = {'type': 'eq', 'fun': lambda x: sum(result) - sum(x)}
+        bounds = tuple((0, sum(result)) for _ in result)
+        res = scipy.optimize.minimize(
+            func, result, method='SLSQP', constraints=constraints, bounds=bounds
+        )
+        return res.x
```

Comment:
I'm slightly worried that this warning is not enough because returning bogus numbers might lead to hard to debug errors. Is there a reason why you don't want to raise or return Optional if optimization fails? You could also add a method parameter raise_on_failure or something like that which could control this behavior (I would set it to True by default)?

Reply:
Raised an error instead of logging a warning.